pkg/controller/cluster/main-controller.go

```diff
@@ -344,143 +344,160 @@ func NewController(podName string, namespacesToWatch set.StringSet, kubeClientSe
+	upgradeServerChannel := make(chan error)
 
-		go func() {
-			klog.Infof("Starting HTTP Upgrade Tenant Image server")
-			wg.Done()
-			if err := c.us.ListenAndServe(); err != http.ErrServerClosed {
-				klog.Infof("HTTP Upgrade Tenant Image server ListenAndServe failed: %v", err)
-				panic(err)
-			}
-		}()
+	// defer close the channels, when this leaderRun exits we close the channel as well
+	defer close(apiServerChannel)
+	defer close(upgradeServerChannel)
 
-		if isOperatorConsoleTLS() {
-			// we need to make sure has console TLS certificate (if enabled)
-			klog.Info("Waiting for Console TLS")
-			wg.Add(1)
-			go func() {
-				defer wg.Done()
-				klog.Infof("Console TLS enabled, starting console TLS certificate setup")
-				err := c.recreateOperatorConsoleCertsIfRequired(ctx)
-				if err != nil {
-					panic(err)
-				}
-				klog.Infof("Restarting Console pods")
-				err = c.rolloutRestartDeployment(getConsoleDeploymentName())
-				if err != nil {
-					klog.Errorf("Console deployment didn't restart: %s", err)
-				}
-			}()
-		} else {
-			klog.Infof("Console TLS is not enabled")
-		}
+	// this wait group only blocks this process until some routines are done
+	var wg sync.WaitGroup
 
-		if IsSTSEnabled() {
-			// Wait for STS API to be ready before starting operator
-			wg.Add(1)
-			go func() {
-				defer wg.Done()
-				klog.Infof("STS is enabled, starting STS API certificate setup")
-				c.generateSTSTLSCert()
-			}()
-		}
+	klog.Info("Waiting for API to start")
+	go startAPIServer(ctx, c, apiServerChannel)
 
-		wg.Wait()
+	klog.Info("Waiting for Upgrade Server to start")
+	go startUpgradeServer(ctx, c, upgradeServerChannel)
 
-		// Start the informer factories to begin populating the informer caches
-		klog.Info("Starting Tenant controller")
+	// 1) we need to make sure we have console TLS certificates (if enabled)
+	if isOperatorConsoleTLS() {
+		klog.Info("Waiting for Console TLS")
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			klog.Infof("Console TLS is enabled, starting console TLS certificate setup")
+			err := c.recreateOperatorConsoleCertsIfRequired(ctx)
+			if err != nil {
+				panic(err)
+			}
+			klog.Infof("Restarting Console pods")
+			err = c.rolloutRestartDeployment(getConsoleDeploymentName())
+			if err != nil {
+				klog.Errorf("Console deployment didn't restart: %s", err)
+			}
+		}()
+	} else {
+		klog.Infof("Console TLS is not enabled")
+	}
 
-		// Wait for the caches to be synced before starting workers
-		klog.Info("Waiting for informer caches to sync")
-		if ok := cache.WaitForCacheSync(stopCh, c.statefulSetListerSynced, c.deploymentListerSynced, c.tenantsSynced, c.policyBindingListerSynced); !ok {
-			panic("failed to wait for caches to sync")
-		}
+	// 2) we need to make sure we have STS API certificates (if enabled)
+	if IsSTSEnabled() {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			klog.Infof("STS is enabled, starting STS API certificate setup")
+			c.generateSTSTLSCert()
```

Comment:
can we start the controller without STS, and have STS start whenever it's ready in it's own goroutine?

Reply:
It does, here inside the `leaderRun` we only wait for the leader to generate the TLS certificate, leader and no leaders start STS server and wait for the certificate to be ready in a no blocking go routine

https://github.com/pjuarezd/operator/blob/0003ab46fa7959bc66a610221fcd39c111f62107/pkg/controller/cluster/main-controller.go#L545-L551

Reply:
but looking at it, is not needed to block the routine to get the certificates here, we could do it right after cache sync, workers and status monitor started, as the last step of the `leaderRun` func.

Reply:
also removed waitgroup